File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/view/ViewPagerPageSelectedObservable.java
package com.jakewharton.rxbinding2.support.v4.view;

import android.support.v4.view.ViewPager;
import android.support.v4.view.ViewPager.OnPageChangeListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

final class ViewPagerPageSelectedObservable extends InitialValueObservable<Integer> {
  private final ViewPager view;

  ViewPagerPageSelectedObservable(ViewPager view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super Integer> observer) {
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnPageChangeListener(listener);
  }

  @Override protected Integer getInitialValue() {
    return view.getCurrentItem();
  }

  static final class Listener extends MainThreadDisposable implements OnPageChangeListener {
    private final ViewPager view;
    private final Observer<? super Integer> observer;

    Listener(ViewPager view, Observer<? super Integer> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override public void onPageSelected(int position) {
      if (!isDisposed()) {
        observer.onNext(position);
      }
    }

    @Override public void onPageScrollStateChanged(int state) {

    }

    @Override protected void onDispose() {
      view.removeOnPageChangeListener(this);
    }
  }
}

File: rxbinding-viewpager/src/main/java/com/jakewharton/rxbinding3/viewpager/ViewPagerPageSelectedObservable.kt
@file:JvmName("RxViewPager")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.viewpager

import androidx.annotation.CheckResult
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.jakewharton.rxbinding2.InitialValueObservable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable of page selected events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 *  *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun ViewPager.pageSelections(): InitialValueObservable<Int> {
  return ViewPagerPageSelectedObservable(this)
}

private class ViewPagerPageSelectedObservable(
  private val view: ViewPager
) : InitialValueObservable<Int>() {

  override fun subscribeListener(observer: Observer<in Int>) {
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnPageChangeListener(listener)
  }

  override val initialValue get() = view.currentItem

  private class Listener(
    private val view: ViewPager,
    private val observer: Observer<in Int>
  ) : MainThreadDisposable(), OnPageChangeListener {

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }

    override fun onPageSelected(position: Int) {
      if (!isDisposed) {
        observer.onNext(position)
      }
    }

    override fun onPageScrollStateChanged(state: Int) {
    }

    override fun onDispose() {
      view.removeOnPageChangeListener(this)
    }
  }
}
